core/src/price_estimation/clients/generic_client.rs

```diff
@@ -1,79 +1,111 @@
-mod api;
-
-use super::{PriceSource, TokenData};
+use super::super::{PriceSource, TokenData};
 use crate::http::HttpFactory;
 use crate::models::TokenId;
 use anyhow::{anyhow, Context, Result};
-use api::{DexagApi, DexagHttpApi};
 use futures::{
     future::{self, BoxFuture, FutureExt as _},
     lock::Mutex,
 };
 use std::collections::HashMap;
 
-struct ApiTokens {
+/// Provides a generic interface to communicate in a standardized way
+/// with specific API token implementations
+pub trait GenericToken {
+    /// Symbol describing the ERC20 token represented by the type instance
+    fn symbol(&self) -> &str;
+}
+
+#[cfg(test)]
+#[derive(Clone, Debug, PartialEq)]
+// Cannot autogenerate with Mockall since the derived traits are needed
+// for testing. GenericToken is a trait that is assigned to the internal
+// token representation of the price source, so the output of `.symbol()`
+// isn't expected to change.
+pub struct MockGenericToken(String);
+#[cfg(test)]
+impl GenericToken for MockGenericToken {
+    fn symbol(&self) -> &str {
+        &self.0
+    }
+}
+#[cfg(test)]
+impl From<&str> for MockGenericToken {
+    fn from(name: &str) -> Self {
+        MockGenericToken(name.to_string())
+    }
+}
+
+#[cfg_attr(test, mockall::automock(type Token=MockGenericToken;))]
+pub trait Api: Sized {
+    type Token: GenericToken + Sync + Send;
+
+    /// Creates a new HTTP interface for connecting to the API
+    fn bind(http_factory: &HttpFactory) -> Result<Self>;
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Self::Token>>>;
+
+    /// Returns the price of one unit of `from` expressed in `to`.
+    /// For example `get_price("ETH", "DAI")` is ~220.
+    fn get_price<'a>(&'a self, from: &Self::Token, to: &Self::Token) -> BoxFuture<'a, Result<f64>>;
+
+    /// Returns a string representing the reference coin in the stead of OWL for this API
+    /// Could be different from "OWL", e.g., when the API does not offer prices with
+    /// respect to OWL
+    fn stable_coin_symbol() -> String;
```

Comment:
Nit: any reason this is not returning `Self::Token`?

Reply:
To make the function return `Self::Token` we would have to either hardcode all components of the token struct as the output of the function or perform external network calls, since the token cache is currently external to the API.
The second solution is inefficient, the first means that we have to add extra checks to confirm that the reference coin is indeed tracked by the exchange when calling `.stable_coin_symbol()`.
So, overall I don't think this change is worth the extra complexity. It would be a different thing however if the cache was internal to the API.